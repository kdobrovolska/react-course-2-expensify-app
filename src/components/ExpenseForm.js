import React from 'react';
import moment from 'moment';
const now =moment();
console.log(now);
import {SingleDatePicker} from 'react-dates';
//import 'react-dates/lib/css/_datepicker.css'; //go to app.js

export default class ExpenseForm extends React.Component
{
    
    constructor(props)
    {
        
      //  alert('conzstructor');
        super(props);
        this.state={description:'',
            note:'', 
            amount:'',
            createdAt:moment(),
            calendarFocused:false,
            error:''
            };
        
        if(!!props.expense)
        {
          this.state =  {...this.state,...props.expense,createdAt:moment(props.expense.CreateAt)};
           
        }
    }
    
    DescriptionChanged=(e)=>
    {
        
        const description=e.target.value;
      //  alert(description);
        this.setState(()=>({description}));
    }
    NoteChanged=(e)=>
    {
        e.persist();
         this.setState(()=>({note:e.target.value}));
    }
    AmountChanged=(e)=>
    {
        
        const amount=e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}));
        }
    }
    onDateChange=(createdAt) =>{
        if(createdAt)
            this.setState(()=>({createdAt}));
    }
    onFocusChange1=({focused})=>
    {
       this.setState(()=>({calendarFocused:focused}));
    }
    onSubmit=(e)=>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount)
        {
            this.setState(()=>({error:'Please provide description and amount'}));
       
        }else{
            this.setState(()=>({error:''}));
            this.props.submit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10),
                note:this.state.note,
                createdAt:this.state.createdAt.valueOf()
            });
        }
    };
    render()
    {
        let obj=this.state;
     /*   if(!!this.props.expense)
        {
           obj=  {...obj,...this.props.expense,createdAt:moment(this.props.expense.CreateAt)};
        }*/

     //   alert('render');
      return <form className='form' onSubmit={ this.onSubmit}>
       {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input type='text' 
            className='text-input' 
            placeholder='Description' 
            autoFocus
            value={obj.description}
            onChange={this.DescriptionChanged}
            />
        <label id='errDescription' className='err'></label>
        <input type='text' 
            className='text-input' 
            placeholder='Amount' 
            value={obj.amount}
            onChange={this.AmountChanged}
            />
            
         <SingleDatePicker 
            date={obj.createdAt}
            onDateChange={this.onDateChange}
            focused ={this.state.calendarFocused}
            onFocusChange ={this.onFocusChange1}
            numberOfMonths={1}
            isOutsideRange={()=>{ return false;}}
            />
        <p></p>
        <textarea placeholder='Add note' 
            className='textarea'
            value={obj.note}
            onChange={this.NoteChanged}
        ></textarea>
        <div>
            <button className='button1'>Save Expense</button>
        </div>
  
        </form> ;
    }
}