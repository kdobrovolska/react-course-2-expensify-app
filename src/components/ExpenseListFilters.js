import React from 'react';
import {connect} from 'react-redux';
import {START_DATE , END_DATE} from 'react-dates/constants'
import {UpdateTextFilter,SortByAmount,SortByDate,SetStartDate,SetEndDate} from '../actions/filters.js';
import {DateRangePicker} from 'react-dates';
export class ExpenseListFilters extends React.Component
{
    state={
            calendarFocused:null,
            
            };
    onDatesChange=({startDate,endDate})=>{
         this.props.SetStartDate(startDate);
        this.props.SetEndDate(endDate);
    };
    onSortChange=(e)=>{
            if(e.target.value==='amount')
                this.props.SortByAmount();
            else
                this.props.SortByDate();
        };

    onTextChange=(e)=>{
            this.props.UpdateTextFilter(e.target.value);
        };

    onFocusChange=(calendarFocused)=>{
        this.setState(()=>({calendarFocused}));
    }
    render()
    {
       // alert('START_DATE='+START_DATE);
        return <div className='content-container'>
        <h1>Filter: </h1>
        <div className='input-group1'>
            <div className='input-group1__item'>
                 <input className='text-input' 
                        type='text'
                        placeholder='Search expense'
                        value={this.props.filter.text} 
                        onChange={this.onTextChange}  />
           </div>
            <div className='input-group1__item'>
                <select className='select1' 
                       value={this.props.filter.sortBy} 
                       onChange={this.onSortChange}>
                    <option>amount</option>
                    <option>date</option>
                </select>
            </div>
            <div className='input-group1__item'>
                <DateRangePicker
                startDate={this.props.filter.startDate}
                endDate={this.props.filter.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput = {this.state.calendarFocused}
                onFocusChange ={this.onFocusChange}
                isOutsideRange={()=>{ return false;}}
                showClearDates={true}
                 />
            </div>
        </div>
        
       
        </div>
    }
};

const MapStoreToProps=(state)=>
{
    return { filter:state.filters};
}
const MapDispatchToProps =(dispatch)=>{
    return { 
        SetStartDate:(startDate)=>dispatch(SetStartDate(startDate)),
        SetEndDate:(endDate)=>dispatch(SetEndDate(endDate)),
        SortByAmount: ()=>dispatch(SortByAmount),
        SortByDate: ()=>dispatch(SortByDate),
        UpdateTextFilter: (text)=>dispatch(UpdateTextFilter(text))
     }
   } 

export default connect(MapStoreToProps,MapDispatchToProps)(ExpenseListFilters);
