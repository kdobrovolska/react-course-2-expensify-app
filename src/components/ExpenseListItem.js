import React from 'react'
import {connect} from 'react-redux';
import {Link} from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'

//import {Dispatch} from '../app.js'

export const ExpenseListItem=({dispatch ,id,description, createdAt,amount})=>
(
   // alert('createdAt'+createdAt);
   // const mom=moment(createdAt).format('MMM Do, YYYY');
  //  alert('mom='+mom);
  <Link className='list-item' to={'/edit/'+id}>

      <div className='list-item__title'>
          <h3> {description}</h3>
          <span className='list-item__sub-title'>{moment(createdAt).format('MMM Do, YYYY')} </span>
      </div>
    
    
    <h3 className='list-item__data'> {numeral(amount/100).format('$0,0.00')}  </h3>
  
  </Link> 
);


export default connect()(ExpenseListItem);