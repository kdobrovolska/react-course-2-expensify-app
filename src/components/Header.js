import React from 'react';
//import ReactDOM from 'react-dom';
import { NavLink } from 'react-router-dom';
import {connect} from 'react-redux';
import {startLogout} from '../actions/auth'


export const  Header= ({startLogout}) => (
    <header>
      <h1>Expensify app</h1>
      <NavLink to='/edit' activeClassName='is-active'>Edit Expense Page </NavLink>
      <NavLink to='/add' activeClassName='is-active'>Add Expense Page </NavLink>
      <NavLink to='/help' activeClassName='is-active' >Help </NavLink>
      <NavLink to='/' activeClassName='is-active' exact={true}>Dashboard </NavLink>
      <button onClick={startLogout}>Logout</button>
      </header>
  );
  const mapDispatchToProps=(dispatch)=>({
    startLogout: ()=>dispatch(startLogout())
  });
 export default connect(undefined, mapDispatchToProps)(Header);