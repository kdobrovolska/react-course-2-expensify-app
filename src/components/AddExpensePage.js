import React from 'react';

import {connect} from 'react-redux';
import {startAddExpense} from '../actions/expenses.js';
import ExpenseForm from './ExpenseForm';

/* const AddExpensePage = (props) => (
  <div>
    <h1> Add Form:</h1>
        <ExpenseForm 
           submit={(obj)=>  {
            console.log(obj);
            props.dispatch(AddExpense(obj));
            props.history.push('/');
        } } 
      />
  </div>
);
export default connect()(AddExpensePage); */


export class  AddExpensePage extends React.Component {
  onSubmit=(expense)=>{
      this.props.startAddExpense(expense);
      this.props.history.push('/');
  }

  render()
  {  return   <div >
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'> Add Form:</h1>
           </div >
        </div>
        <div className='content-container'>
          <ExpenseForm  submit={ this.onSubmit} />
        </div>
    </div>;
  }
}
const MapDispatchToProps =(dispatch)=>{
 return { startAddExpense:(expense)=>dispatch(startAddExpense(expense))}
}
export default connect(undefined,MapDispatchToProps)(AddExpensePage);