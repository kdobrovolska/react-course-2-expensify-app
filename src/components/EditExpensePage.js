import React from 'react';
import {connect} from 'react-redux';
import moment from 'moment';
import ExpenseForm from './ExpenseForm';
import {startEditExpense, RemoveExpense,startRemoveExpense} from '../actions/expenses.js';

export class EditExpensePage extends React.Component{

   submit=(obj)=>  {
          this.props.startEditExpense(this.props.expense.id,obj);
          this.props.history.push('/');
        }

   onClick=()=>{
          this.props.startRemoveExpense(this.props.expense.id);
          this.props.history.push('/');
        }

    render()
    {
        return <div>
          <div className='page-header'>
            <div className='content-container'>
                <h1 className='page-header__title'>Edit Expense </h1>
            </div>
           </div>
           <div className='content-container'>
              <ExpenseForm expense={this.props.expense}
                submit={this.submit } />
              <button className='button1 button1__secondary' onClick={ this.onClick}>Remove Expense</button>
          </div>
        </div>;
    }
};

const MapStoreToProps=(state,props)=>
{
     return { expense:state.expenses.filter(
       (item)=>(item.id===props.match.params.id))[0]
     }
}

 const MapDispatchToProps =(dispatch)=>{
  return { 
    startEditExpense:(id,expense)=>dispatch(startEditExpense(id,expense)),
    startRemoveExpense: (id)=>dispatch(startRemoveExpense(id))
   }
 } 

export default connect(MapStoreToProps, MapDispatchToProps)(EditExpensePage);
//export default connect(MapStoreToProps)(EditExpensePage);