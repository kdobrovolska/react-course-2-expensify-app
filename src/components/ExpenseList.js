import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem.js'
import VisibleExpenses from '../selectors/expenses.js'

export const ExpenseList=(props)=>
(
    <div className='content-container'>
        <div className='list-header'>
            <div className='show-for-mobile'> Expenses  </div>
            <div className='show-for-desktop'> Expense  </div>
            <div className='show-for-desktop'> Amount  </div>
        </div>
        <div className='list-body'>
        {
           ( props.expenses.length===0)?(
               <div className='list-item list-item__message'>No expenses </div>
           ):(
                    props.expenses.map((item,ident)=>{
                    
                        return <ExpenseListItem key={ident} {...item}/>;
                
                    })
            )
        }
        </div>
    </div>
);
// first variant
//const ConnectedExpenseList=connect((st)=>{
  //  return {
  //      expenses:st.expenses
  //  };
//})(ExpenseList);
//export default ConnectedExpenseList;

const mapStateToProps=(state)=>
{
    return{
        expenses:VisibleExpenses(state.expenses,state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);