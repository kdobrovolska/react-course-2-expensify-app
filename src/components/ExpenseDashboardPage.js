import React from 'react';
import ExpenseList from './ExpenseList.js'
import ExpenseListFilters from './ExpenseListFilters.js'
import ExpensesSummary from './ExpensesSummary'

const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList type='new'/>
    
  </div>
);
export default ExpenseDashboardPage;