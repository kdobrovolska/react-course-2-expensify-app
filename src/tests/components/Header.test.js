import React from 'react';
import { shallow } from 'enzyme';
//import toJSON from 'enzyme-to-json';
//import ReactShallowRenderer from 'react-test-renderer/shallow';
import {Header} from '../../components/Header';

test('should render Header correctly', ()=>{
   const wrapper= shallow(<Header startLogout={()=>{}}/>);
   expect(wrapper).toMatchSnapshot();
    
});


test('should check button click', ()=>{
  const startLogout=jest.fn();
  const wrapper= shallow(<Header startLogout={startLogout}/>);
  wrapper.find('button').at(0).simulate('click');
  expect(startLogout).toHaveBeenCalled();
   
});