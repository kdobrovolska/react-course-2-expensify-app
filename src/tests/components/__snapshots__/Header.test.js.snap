// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__contant"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Expensify
        </h1>
      </Link>
      <button
        className="button1 button1__link"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
